libjolie/src/main/java/jolie/lang/parse/module/Finder.java

```diff
@@ -0,0 +1,235 @@
+		Path olPath = basePath.resolve( filename + ".ol" );
+		if( Files.exists( olPath ) ) {
+			return olPath.toFile();
+		}
+		throw new FileNotFoundException( olPath.toString() );
+	}
+
+	/**
+	 * perform lookup to the target module's
+	 * 
+	 * @param basePath base path for lookup
+	 * @return Source object corresponding to the module.
+	 * @throws FileNotFoundException when a module is not found
+	 */
+	protected Source locateModule( Path basePath ) throws FileNotFoundException {
+		Path packagePath = Paths.get( basePath.toString(), this.packagesToken() );
+		File olTargetFile = this.olLookup( packagePath, this.moduleName() );
+		if( !olTargetFile.exists() ) {
+			throw new FileNotFoundException( olTargetFile.toString() );
+		}
+		return new FileSource( olTargetFile );
+	}
+
+}
+
+
+/**
+ * A class represent the finder for relative ImportStatement
+ */
+class RelativePathFinder extends Finder {
+
+	/**
+	 * an URI source of the caller
+	 */
+	private final URI source;
+	private int packagesTokenStartIndex = 0;
+
+	protected RelativePathFinder( String[] target, URI source ) {
+		super( target );
+		assert source != null;
+		this.source = source;
+	}
+
+
+	/**
+	 * resolve path from source, each dot prefix means 1 level higer from the caller path directory
+	 */
+	private Path resolveDotPrefix() {
+		Path sourcePath = Paths.get( source );
+		Path basePath;
+		if( !sourcePath.toFile().isDirectory() ) {
+			basePath = sourcePath.getParent();
+		} else {
+			basePath = sourcePath;
+		}
+		int i = 1;
+		for( ; i < this.target.length - 1; i++ ) {
+			if( target[ i ].isEmpty() ) {
+				basePath = basePath.getParent();
+			} else {
+				break;
+			}
+		}
+		packagesTokenStartIndex = i;
+		return basePath;
+	}
+
+	@Override
+	public Source find() throws ModuleNotFoundException {
+		Path basePath = resolveDotPrefix();
+		try {
+			Source moduleFile = super.locateModule( basePath );
+			return moduleFile;
+		} catch( FileNotFoundException e ) {
+			throw new ModuleNotFoundException( this.moduleName(), e.getMessage() );
+		}
+	}
+
+	@Override
+	protected String[] packagesToken() {
+		return Arrays.copyOfRange( this.target, packagesTokenStartIndex, this.target.length - 1 );
+	}
+
+}
+
+
+/**
+ * A class represent the finder for absolute path import statement
+ */
+class AbsolutePathFinder extends Finder {
+
+	private final Path[] packagesDirectories;
+	private final Path workingDirectory;
+
+	public AbsolutePathFinder( String[] target, Path workingDirectory, Path[] packagesDirectories ) {
+		super( target );
+		this.packagesDirectories = packagesDirectories;
+		this.workingDirectory = workingDirectory;
+	}
+
+
+	@Override
+	public Source find() throws ModuleNotFoundException {
+		/**
+		 * 1. Try to resolve P directly from WDIR. 2. Check if FIRST.jap is in WDIR/lib. If so, resolve REST
+		 * inside of this jap. 3. Try to resolve P from the list of packages directories.
+		 */
+		ModuleNotFoundException err = new ModuleNotFoundException( this.moduleName() );
```

Comment:
It's weird that you create an exception you might not need. Consider using a list of messages in this method instead.